commit 4a255e2b5ec3f2fb67e3333ed8b47ef43fbd27e3
Author:     Yuhong Wen <yh_wen@hotmail.com>
AuthorDate: Tue Mar 8 13:05:35 2022 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Tue Mar 8 13:05:35 2022 -0500

    Ha support (#261)
    
    * Cifar10 scaffold (#127)
    
    * SCAFFOLD Integration
    
    scaffold learner
    
    scaffold configs
    
    initialize the scaffold terms on the client
    
    update README to include SCAFFOLD description
    
    add scaffold experiment
    
    update urls
    
    add LICENSE from NIID-Bench
    
    update README and plots to reflect SCAFFOLD experiment
    
    refactor scaffold computation steps into their own member functions
    
    remove unused import
    
    refactor to use ScaffoldLearner class
    
    Scaffold learner depends on PyTorch. Renamed as such
    
    use two standard aggregators, inherit SCAFFOLD workflow from standard ScatterAndGather
    
    remove custom aggregator
    
    use update_shareable
    
    use multi-class inheritance for scaffold learner
    
    use new aggregator that supports several DXOs
    
    fix updating call for global model
    
    simplify dxo/shareable handling
    
    use built-in PTFileModelLocator
    
    fix formatting
    
    add PT Formatter
    
    use built-in validation json generator; remove custom formatter
    
    remove custom Json validation generator
    
    update license
    
    restore run_secure.sh
    
    restore project yml
    
    restore main branch learner executor
    
    add SCAFFOLD link to example readme
    
    remove custom validation json generator
    
    print model owner info during validation
    
    remove custom formatter
    
    remove special handling of validation dxo. Not needed anymore
    
    use zeros_like() to initialize scaffold terms
    
    * use scaffold helper class
    
    * formatting
    
    * make scaffold function args consistent
    
    * Add versioneer so nvflare.__version__ has meaningful value (#215)
    
    * Add versioneer so nvflare.__version__ has meaningful value
    
    * Fix coding style and license for versioneer
    
    * Add information about cross site validation, commands, and hostname to documentation (#206)
    
    * Add information about cross site validation to documentation
    
    * More minor documentation updates
    
    * Fix link
    
    * Add persistence to admin cli (#210)
    
    * Add persistence to admin cli
    
    * save to ~/.nvflare
    
    * 211 Update hello-monai example (#212)
    
    * update hello-monai
    
    Signed-off-by: Yiheng Wang <vennw@nvidia.com>
    
    * decrease aggregation_epochs
    
    Signed-off-by: Yiheng Wang <vennw@nvidia.com>
    
    * Initial implementation of Ditto algorithm (#144)
    
    Refactor ditto with multi inheritance
    
    Refactor ditto with multi inheritance
    
    Update README.md
    
    move learners to app_common with other minor updates
    
    update license section
    
    update for unit test
    
    update for unit test
    
    update for unit test
    
    remove redundant init parameters
    
    Update supervised_ditto_learner.py
    
    Update README.md
    
    Update README.md
    
    Update README.md
    
    Update README.md
    
    Update supervised_ditto_learner.py
    
    move learners back to custom folder, will further discuss and make the move in separate PR
    
    * Initial implementation for HA.
    
    * Added FL client HA support.
    
    * Enhanced the Client HA.
    
    * HA function working for server & client.
    
    * Working with SnapshotFilePersistor config.
    
    * Made the overseer_agent configurable.
    
    * Made the admin support HA.
    
    * Adjust retry.
    
    * Adjust admin prompt.
    
    * Added storage_state_persistor.py
    
    * Integrated with new overseer-agent.
    
    * Changed to use admin Overseer-agent.
    
    * Updated overseer_spec.py
    
    * Fixed end line.
    
    * Added storage and used the dev-2.1 cli.py.
    
    * Used dev-2.1 cli.py.
    
    Co-authored-by: Holger Roth <6304754+holgerroth@users.noreply.github.com>
    Co-authored-by: Isaac Yang <isaacy@nvidia.com>
    Co-authored-by: nvkevlu <55759229+nvkevlu@users.noreply.github.com>
    Co-authored-by: Sean Yang <seany314@gmail.com>
    Co-authored-by: Yiheng Wang <68361391+yiheng-wang-nv@users.noreply.github.com>
    Co-authored-by: Ziyue Xu <71786575+ZiyueXu77@users.noreply.github.com>
---
 nvflare/private/fed/client/client_engine.py |  2 +-
 nvflare/private/fed/server/server_engine.py | 48 ++++++++++++++++++++++++-----
 2 files changed, 42 insertions(+), 8 deletions(-)

diff --git a/nvflare/private/fed/client/client_engine.py b/nvflare/private/fed/client/client_engine.py
index 210fdd38..df3dbf12 100644
--- a/nvflare/private/fed/client/client_engine.py
+++ b/nvflare/private/fed/client/client_engine.py
@@ -144,7 +144,7 @@ class ClientEngine(ClientEngineInternalSpec):
         if os.path.exists(token_file):
             os.remove(token_file)
         with open(token_file, "wt") as f:
-            f.write("%s\n%s\n%s\n%s\n" % (self.client.token, run_number, self.client.client_name, open_port))
+            f.write("%s\n%s\n%s\n%s\n%s\n" % (self.client.token, self.client.ssid, run_number, self.client.client_name, open_port))
 
     def wait_process_complete(self):
         self.client.process.wait()
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index 3c5d0837..aae3b3bd 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import copy
 import gc
 import logging
 import os
@@ -26,16 +27,17 @@ from threading import Lock
 from typing import List, Tuple
 
 from nvflare.apis.client import Client
-from nvflare.apis.fl_constant import MachineStatus, ReservedTopic, ReturnCode
+from nvflare.apis.fl_constant import MachineStatus, ReservedTopic, ReturnCode, FLContextKey, SnapshotKey
 from nvflare.apis.fl_context import FLContext
+from nvflare.apis.fl_snapshot import FLSnapshot
 from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.utils.fl_context_utils import get_serializable_data
 from nvflare.apis.workspace import Workspace
 from nvflare.fuel.hci.zip_utils import zip_directory_to_bytes
 from nvflare.private.admin_defs import Message
 from nvflare.private.fed.server.server_json_config import ServerJsonConfigurator
 from nvflare.widgets.info_collector import InfoCollector
 from nvflare.widgets.widget import Widget, WidgetID
-
 from .client_manager import ClientManager
 from .run_manager import RunManager
 from .server_engine_internal_spec import EngineInfo, RunInfo, ServerEngineInternalSpec
@@ -43,7 +45,7 @@ from .server_status import ServerStatus
 
 
 class ServerEngine(ServerEngineInternalSpec):
-    def __init__(self, server, args, client_manager: ClientManager, workers=3):
+    def __init__(self, server, args, client_manager: ClientManager, snapshot_persistor, workers=3):
         """Server engine.
 
         Args:
@@ -76,6 +78,7 @@ class ServerEngine(ServerEngineInternalSpec):
         self.logger = logging.getLogger(self.__class__.__name__)
 
         self.asked_to_stop = False
+        self.snapshot_persistor = snapshot_persistor
 
     def _get_server_app_folder(self):
         return "app_server"
@@ -134,7 +137,7 @@ class ServerEngine(ServerEngineInternalSpec):
     def validate_clients(self, client_names: List[str]) -> Tuple[List[Client], List[str]]:
         return self._get_all_clients_from_inputs(client_names)
 
-    def start_app_on_server(self) -> str:
+    def start_app_on_server(self, snapshot=None) -> str:
         if self.run_number == -1:
             return "Please set a run number."
 
@@ -157,7 +160,7 @@ class ServerEngine(ServerEngineInternalSpec):
 
             # future = self.executor.submit(start_server_training, (self.server))
             future = self.executor.submit(
-                lambda p: start_server_training(*p), [self.server, self.args, app_root, self.run_number]
+                lambda p: start_server_training(*p), [self.server, self.args, app_root, self.run_number, snapshot]
             )
 
             start = time.time()
@@ -395,6 +398,37 @@ class ServerEngine(ServerEngineInternalSpec):
 
         return results
 
+    def persist_components(self, fl_ctx: FLContext, completed: bool):
+
+        # Call the State Persistor to persist all the component states
+        # 1. call every component to generate the component states data
+        #    Make sure to include the current round number
+        # 2. call persistence API to save the component states
+
+        snapshot = FLSnapshot()
+        for component_id, component in self.run_manager.components.items():
+            snapshot.save_component_snapshot(component_id=component_id,
+                                             component_state=component.get_persist_state(fl_ctx))
+
+        snapshot.save_component_snapshot(component_id=SnapshotKey.FL_CONTEXT,
+                                         component_state=copy.deepcopy(get_serializable_data(fl_ctx).props))
+
+        workspace = fl_ctx.get_prop(FLContextKey.WORKSPACE_OBJECT)
+        data = zip_directory_to_bytes(workspace.get_run_dir(self.run_number), "")
+        snapshot.save_component_snapshot(component_id=SnapshotKey.WORKSPACE,
+                                         component_state={"content": data})
+
+        snapshot.completed = completed
+
+        self.server.snapshot_location = self.snapshot_persistor.save(snapshot=snapshot)
+        self.logger.info(f"persist the snapshot to: {self.server.snapshot_location}")
+
+    def restore_components(self, snapshot: FLSnapshot, fl_ctx: FLContext):
+        for component_id, component in self.run_manager.components.items():
+            component.restore(snapshot.get_component_snapshot(component_id=component_id), fl_ctx)
+
+        fl_ctx.props.update(snapshot.get_component_snapshot(component_id=SnapshotKey.FL_CONTEXT))
+
     def dispatch(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareable:
         return self.run_manager.aux_runner.dispatch(topic=topic, request=request, fl_ctx=fl_ctx)
 
@@ -402,7 +436,7 @@ class ServerEngine(ServerEngineInternalSpec):
         self.executor.shutdown()
 
 
-def start_server_training(server, args, app_root, run_number):
+def start_server_training(server, args, app_root, run_number, snapshot):
 
     restart_file = os.path.join(args.workspace, "restart.fl")
     if os.path.exists(restart_file):
@@ -418,7 +452,7 @@ def start_server_training(server, args, app_root, run_number):
 
         set_up_run_config(server, conf)
 
-        server.start_run(run_number, app_root, conf, args)
+        server.start_run(run_number, app_root, conf, args, snapshot)
     except BaseException as e:
         traceback.print_exc()
         logging.getLogger().warning("FL server execution exception: " + str(e))
